### PR TITLE
Refactoring the way data is returned in panedr

### PR DESCRIPTION
**Pull request by BFedder**
Addressing #25 

I have changed the way data than panedr returns the energy data that it reads from EDR files. 

### Old behaviour
`edr_to_df()` opens the file, parses its binary contents, populates lists with the energy data, and assembles and returns a Pandas DataFrame. 

### New behaviour
The parsing of the EDR file is now done by a new `read_edr()` function. 
`read_edr()` returns the lists that were previously generated by `edr_to_df()` as an intermediate step.
The data can now be returned as a dictionary of NumPy arrays with `edr_to_dict()`. Alternatively, returning the data as a Pandas DataFrame is possible via `edr_to_df()`. Both of these functions call `read_edr()`.

The old behaviour is maintained, users can still call `edr_to_df()` and obtain the results they expect.

### To Do
- [x] Write tests for new functions
- [ ] Write detailed documentation for new functions / panedr in general (#30)
- [ ] Include more type hinting (#31)

### CODE REVIEW

File: panedr/panedr.py [hunk 1]
```diff
@@ -395,21 +395,21 @@ def edr_strings(data, file_version, n):
...
-    
+
     Does not roll the reading position back.
     """
     magic = data.unpack_int()
     return magic == -7777777
 
 
-def edr_to_df(path, verbose=False):
+def read_edr(path, verbose_set=False):
```
Comment (hmacdope) [inline panedr/panedr.py:405]: Why the change from `verbose`?

File: panedr/panedr.py [hunk 2]
```diff
@@ -421,11 +421,28 @@ def edr_to_df(path, verbose=False):
...
               .format(ifr, frame.t),
               end='', file=sys.stderr)
         print('\n{} frame read in {:.2f} seconds'.format(ifr, end - begin),
               file=sys.stderr)
+
+    return all_energies, all_names, times
+
+
+def edr_to_df(path: str, verbose: bool = False):
+    import pandas
```
Comment (hmacdope) [inline panedr/panedr.py:435]: Is pandas now an optional dependency in requirements.txt etc? If so I would guard this with a `try: except` similar to 

```python
        try:
            import pandas
        except ImportError:
            raise ImportError("""ERROR --- pandas was not found!
                pandas is required to use the `.edr_to_df()` functionality.
                try installing it using pip eg:
                    pip install pandas """)

```

File: panedr/panedr.py [hunk 3]
```diff
@@ -421,11 +421,28 @@ def edr_to_df(path, verbose=False):
...
+    all_energies, all_names, times = read_edr(path, verbose_set=verbose)
     df = pandas.DataFrame(all_energies, columns=all_names, index=times)
     return df
+
+
+def edr_to_dict(path: str, verbose: bool = False):
+    import numpy as np
```
Comment (hmacdope) [inline panedr/panedr.py:442]: Is this an optional dependency? I think its probably safe to make `numpy` compulsory for panedr.  If we do make it compulsory you can import this at the top of the file otherwise use a guard like the one for pandas mentioned in my other comment. 

Thoughts @jbarnoud?

File: tests/test_edr.py [hunk 4]
```diff
@@ -163,6 +163,14 @@ def _assert_progress_range(self, progress, dt, start, stop, step):
             assert ref_line == progress_line
 
 
+def test_edr_to_dict():
```
Comment (hmacdope) [inline tests/test_edr.py:166]: This only really tests that it returns the same as `edr_to_df` so perhaps a more descriptive name like `edr_to_dict_matches_edr_to_df`. I know its a bit verbose but clarity is always good.

**Review by hmacdope — CHANGES_REQUESTED** [review]
Good start. have a look at comment for some changes.

File: panedr/panedr.py (hunk 3 above)
Comment (jbarnoud) [reply panedr/panedr.py:442]: I think it is reasonable to make numpy a compulsory dependency. The current users have to install it already because of pandas anyway; the new users will likely use it as well.

File: panedr/panedr.py [hunk 5]
```diff
@@ -75,7 +75,7 @@
 Enxnm = collections.namedtuple('Enxnm', 'name unit')
 ENX_VERSION = 5
 
-__all__ = ['edr_to_df']
+__all__ = ['edr_to_df', 'edr_to_dict']
```
Comment (jbarnoud) [inline panedr/panedr.py:78]: I don't see a reason not to add `read_edr` here as well.

File: panedr/panedr.py [hunk 6]
```diff
@@ -421,11 +421,28 @@ def edr_to_df(path, verbose=False):
...
+
+
+def edr_to_dict(path: str, verbose: bool = False):
+    import numpy as np
+    all_energies, all_names, times = read_edr(path, verbose_set=verbose)
+    energy_dict = {}
+    for idx, name in enumerate(all_names):
+        energy_dict[name] = np.array(
+            [all_energies[frame][idx] for frame in range(len(times))])
+    return energy_dict
```
Comment (jbarnoud) [inline panedr/panedr.py:448]: Make sure that the "Time" key is in. I expect it to be, but I do not remember exactly how I treated it.

**Review by jbarnoud — COMMENTED** [review]
I see no issue with the tests depending on pandas. Optional tests are usually a bad idea because they allow issues to fall through the cracks. It is justifiable in some cases, though, like when a dependency is very specialised and difficult to install.

File: panedr/panedr.py (hunk 1 above)
Comment (BFedder) [reply panedr/panedr.py:405]: I felt weird about `verbose=verbose`, but yeah, I guess there is not really a reason not to put this :sweat_smile:

File: panedr/panedr.py (hunk 2 above)
Comment (BFedder) [reply panedr/panedr.py:435]: Good point. I'll make a note to add a test for raising this error as well.

File: panedr/panedr.py (hunk 5 above)
Comment (BFedder) [reply panedr/panedr.py:78]: I saw `read_edr` as merely providing data for the two user-exposed functions `edr_to_df` and `edr_to_dict`, so that the user should never need to call `read_edr` itself directly. I am not sure if the return values of this function are of use to a user, but yeah, that's not really a reason to add it here.

File: panedr/panedr.py (hunk 6 above)
Comment (BFedder) [reply panedr/panedr.py:448]: "Time" is part of `all_names` and `all_energies` as returned by `read_edr`. This is done specifically by 

```python
all_names = [u'Time'] + [nm.name for nm in edr_file.nms]
[...]
all_energies.append([frame.t] + [ener.e for ener in frame.ener])
```

File: panedr/panedr.py (hunk 2 above)
Comment (BFedder) [reply panedr/panedr.py]: I wasn't sure what the best way to make pandas optional is. I have now done this by removing pandas from requirements.txt and adding a section under [extras] in setup.cfg. panedr can now be installed with pandas by running 

`pip install -e .[pandas]`

Please let me know if this is not the best way to do these things.

Comment (jbarnoud) [reply panedr/panedr.py]: I would rather do things a little bit differently. The way you do it here will break the user experience for the standalone case. A user who do not use mdanalysis will install panedr and won't have pandas to use the main function out of the box.

Instead, I would create 2 packages: the default one that depends on pandas and a "lite" one for downstream integrators who want to minimise dependencies.

I don't know how to do that, though...

File: panedr/panedr.py [hunk 7]
```diff
@@ -427,5 +428,27 @@ def edr_to_df(path, verbose=False):
...
               file=sys.stderr)
+
+    return all_energies, all_names, times
+
+
+def edr_to_df(path: str, verbose: bool = False):
+    try:
+        import pandas
+    except ImportError:
+        raise ImportError("""ERROR --- pandas was not found!
```
Comment (jbarnoud) [inline panedr/panedr.py:439]: There should be a test for that, which means a pipeline that does not install pandas and a test that runs the function and asserts the exception is raised.

File: requirements.txt [hunk 8]
```diff
@@ -1,2 +1,2 @@
-pandas
+numpy>=1.19.0
```
Comment (jbarnoud) [inline requirements.txt:1]: @IAlibay is bumping the version to 1.20.0 for MDAnalysis in https://github.com/MDAnalysis/mdanalysis/pull/3737

Comment (IAlibay) [reply requirements.txt:1]: Doesn't impact things here so it wouldn't think it matters. If you want to raise numpy to 1.20 you'll have to drop py3.6.

File: panedr/panedr.py (hunk 7 above)
Comment (IAlibay) [reply panedr/panedr.py:439]: @BFedder you should be able to just use monkeypatch here to do this: https://github.com/MDAnalysis/mdanalysis/blob/3e249fe7173e68bb61d4ad7a2dfd316014760410/testsuite/MDAnalysisTests/utils/test_datafiles.py#L28-L38

File: requirements.txt (hunk 8 above)
Comment (BFedder) [reply requirements.txt:1]: What should we prioritise here? Staying in-step with MDAnalysis' dependencies or keeping Py3.6?

**Comment by BFedder** [comment]
I'm thinking if there are no further comments here, it might make sense to merge this PR as is so that #42 can be merged soon as well. I will then add docstrings/type hints in a separate PR. What do you think @jbarnoud @hmacdope @IAlibay?

File: requirements.txt (hunk 8 above)
Comment (IAlibay) [reply requirements.txt:1]: It doesn't impact MDAnalysis if panedr supports a wider range of python & numpy versions, so IMHO it's fine to just leave it as-is.

File: tests/test_edr.py [hunk 9]
```diff
@@ -59,6 +59,20 @@
...
                                    'xvgprec', 'edrfile', 'xvgfile'])
 
+
+def test_failed_import(monkeypatch):
+    # Putting this test first to avoid datafiles already being loaded
+    errmsg = "ERROR --- pandas was not found!"
+
+    monkeypatch.setitem(sys.modules, 'pandas', None)
+
+    if 'MDAnalysis.tests.datafiles' in sys.modules:
```
[the remark this answers is not in the document]
Comment (BFedder) [reply tests/test_edr.py]: Ooops... Thanks

File: tests/test_edr.py [hunk 10]
```diff
@@ -163,10 +172,21 @@ def _assert_progress_range(self, progress, dt, start, stop, step):
...
+def test_edr_to_dict_matches_edr_to_df():
+    array_dict = panedr.edr_to_dict(EDR)
+    ref_df = panedr.edr_to_df(EDR)
+    array_df = pandas.DataFrame.from_dict(array_dict).set_index(
+        "Time", drop=False)
+    assert array_df.equals(ref_df)
+
+
+# TODO: write test that ImportError is raised when pandas not installed
+
```
Comment (IAlibay) [inline tests/test_edr.py:184]:
```suggestion
```
This is done now right?

**Review by IAlibay — CHANGES_REQUESTED** [review]
Just the one thing on my end - I'm happy to move "further cleaning up" (docstring, type hints) to a future PR if this means we can keep things progressing.

@hmacdope can you check if everything you requested has been addressed?

**Review by hmacdope — APPROVED** [review]
All looks good to me @IAlibay.  Great work @BFedder keep it up!

File: tests/test_edr.py (hunk 10 above)
Comment (hmacdope) [reply tests/test_edr.py:184]: I can merge once you clarify this @BFedder

Comment (BFedder) [reply tests/test_edr.py]: Yes indeed, this test is now present. I removed the TODO
